docs/network-drivers.md

```diff
@@ -1,47 +1,127 @@
-should be set: ``trunk_port`` is the name of the port that headnode traffic
-runs on.  This port is always on all HaaS VLANs, so the deployment tests need
-to know this port's identity in order to check that the switch is configured
-correctly.
-
-# Multiple-switch  driver using VLAN isolation (``complex_vlan``)
-
-There is one configuration field: ``switches`` should be a JSON list of
-objects, each representing one physical switch.  Each object must have a field
-``switch``, the switch driver to be used, and ``name``, a unique identifier
-that is used when specifying ports.  The remaining fields are used by the
-switch driver.
-
-When using this driver, name ports in the format ``[switch]::[port]``, where
-``[switch]`` is the ``name`` field of the appropriate switch, and ``[port]``
-is the port on the switch according to that switch's driver.  For example,
-port ``gi1/1/15`` on switch ``dell-15`` would be named ``dell-15::gi1/1/15``.
-There is not currently any provision for escaping a ``::`` in a switch or port
-name.
-
-When runnning deployment tests, ``deployment.cfg`` has an additional value
-``trunk_ports``, a JSON-encoded list of port names.  This should contain all
-ports that are set to trunk HaaS network traffic, either between switches or
-to the HaaS headnode host.
+This document describes the network driver model, primarily from a
+system administrator's standpoint, though there are some details useful
+for developers writing tools which interact with the API as well.
+
+# Overview
+
+There are two classes of network-related drivers: network allocators and
+switches. Switches are straightforward to understand; they provide
+support for a particular hardware switch. Network allocators are a
+bit more abstract.
+
+HaaS networks may be backed by various underlying isolation technologies
+(currently the only network allocator shipped with HaaS uses 802.1q
+VLANs, but there may be more, e.g. VXLANs in the future). A network
+allocator manages the details of mapping networks created by HaaS to
+these underlying technologies. For example, the ``vlan_pool`` allocator
+maps each network to a unique VLAN id.
+
+Drivers are implemented as extensions, and must be added to the
+``[extensions]`` section of ``haas.cfg``. You must supply *exactly* one
+network allocator driver, and one or more switches. For example::
+
+    ...
+    [extensions]
+    haas.ext.network_allocators.vlan_pool =
+    haas.ext.switches.dell =
+    haas.ext.switches.nexus =
+
+Some drivers may also need driver specific options, which go in a
+section with the same name as the extension, e.g.::
+
+    [haas.ext.network_allocators.vlan_pool]
+    vlans = 300-500, 700-750
+
+# Network allocator drivers
+
+The only network allocator shipped with HaaS that is of interest to
+users (there are others useful for development purposes) is the VLAN
+allocator. The name of the extension is
+``haas.ext.network_allocators.vlan_pool``, and it requires a single
```

Comment:
Have you given thought to how we'll handle divergent vlan_pools managed by a single HaaS instance (like I'm told junipers can work)? Configure separate vlan_pools via the API?